feat/keyword-create

backswan0 opened the pull request:
## #️⃣ Kan Board Number

18

## 📝 요약

- 사용자 채용 키워드 알림 등록 (C) 구현 완료 - Postman 테스트 완료 
- 존재하지 않는 키워드일 때 not found 예외 처리 
- 사용자가 키워드의 id를 입력하면 중간테이블인 user_keyword에 저장되도록 로직 작성 

## 🛠️ PR 유형

- [x] 새로운 기능 추가
- [x] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 논의사항 or 질문

등록이 완료되었다고 메시지만 보내도 괜찮을까요? 지금은 등록된 키워드의 식별자를 반환하는데 '굳이 또 알려줘야 할까?' 의문이 살짝 들어서요:) 

## ✅ PR Checklist

PR이 다음 요구 사항을 충족했는지 확인하기!

- [x] 커밋 메시지 컨벤션 준수
- [x] 변경 사항 테스트 여부 체크(버그 수정/기능 테스트)

## 💥 리뷰어 지목!!!!!

<!--- 리뷰어를 2명 이상 지목해주세요. -->
- [x] 채영
- [ ] 지현
- [x] 리은
- [x] 승찬
- [ ] 주양

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **새로운 기능**
  - 사용자 키워드 관리 기능이 추가되어, 사용자가 선택한 키워드 목록을 등록하고 확인할 수 있습니다.
  - 제출한 키워드의 유효성이 자동으로 검증되며, 존재하지 않는 키워드를 선택할 경우 "존재하지 않는 키워드입니다."라는 오류 메시지가 표시됩니다.
  - 등록 성공 시 선택한 키워드 목록이 사용자에게 반환되어, 보다 원활한 경험을 제공합니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

backswan0 committed:
:tada: add: 요청 및 응답 DTO 생성 및 엔티티에 정적 팩토리 메서드 추가

- 도메인: UserKeyword

- 사용자 채용 키워드 알림 등록용 요청 및 응답 DTO 생성

- 엔티티 내부에 팀 컨벤션에 맞게 정적 팩토리 메서드 추가

backswan0 committed:
:sparkles: feat: 사용자 채용 키워드 알림 등록 기능 구현

- [필요] dev git pull 이후 사용자 정보 가져오는 로직 수정 및 테스트 예정

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into feat/keyword-create

backswan0 committed:
:recycle: refactor: 사용자 채용 키워드 알림 등록 기능에 예외 처리 추가

- 도메인: UserKeyword

- 사용자가 입력한 키워드의 id가 없으면 예외 발생하도록 리팩토링

- ErrorCode에 존재하지 않는 키워드라는 404 예외 메시지 추가

- 컨트롤러 및 서비스에 Slf4j 어노테이션 추가

- [할 일] AOP로 Slf4j 어노테이션 대체

- [할 일] 유저 불러오는 로직 완료되면 UserKeywordService 리팩토링

backswan0 committed:
:recycle: refactor: 요청 DTO 안에 리스트로 받을 수 있도록 리팩토링

- 키워드 등록 시 배열로 한꺼번에 키워드 식별자를 받을 수 있도록 리팩토링

- [할 일] AOP로 Slf4j 어노테이션 대체

- [할 일] 유저 불러오는 로직 완료되면 UserKeywordService 리팩토링

backswan0 committed:
:memo: chore: 서비스 레이어에서 충돌하는 예외 메시지 우선 삭제

backswan0 committed:
Merge branch 'dev' of https://github.com/cheer-ha/cheer-ha into feat/keyword-create

# Conflicts:
#	src/main/java/com/project/cheerha/common/exception/ErrorCode.java

backswan0 committed:
:recycle: refactor: 요청 및 응답 DTO와 controller 및 service 리팩토링

- 여러 DTO가 아니라 한 DTO 안에 리스트 형태로 키워드 식별자를 주고 받을 수 있도록 수정

- Auth 어노테이션을 사용하여 유저의 id 가져옴

- [할 일] 키워드가 없을 때 예외 처리 추가

- [할 일] 사용자가 이미 등록한 키워드 id를 재차 입력했을 때 예외 처리 해야 할 듯

backswan0 committed:
:sparkles: feat: 예외 처리 추가

- id로 키워드를 조회할 수 없을 때 404 예외 처리 및 메시지 추가

- 서비스 레이어에 예외 처리

- [할 일] 이미 등록된 키워드를 또 등록하려고 할 때 예외 처리

roqkfchqh commented on src/main/java/com/project/cheerha/domain/keyword/controller/UserKeywordController.java:
```diff
@@ -9,4 +18,20 @@
+    private final UserKeywordService userKeywordService;
+
+    @PostMapping("/user/keywords")
+    public ResponseEntity<CreateUserKeywordResponseDto> registerUserKeywordList(
```
createUserKeywordList로 변경해주세요

backswan0 replied:
넵, 수정했습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/keyword/controller/UserKeywordController.java:
```diff
@@ -9,4 +18,20 @@
+        @RequestBody CreateUserKeywordRequestDto requestDto,
+        @Auth AuthUser authUser
+    ) {
+        Long userId = authUser.id();
```
해당 로직은 서비스 계층에 존재하는 것이 더 좋을 것 같습니다

roqkfchqh replied:
userId만 정제해서 보내기보단 전체를 보내는 것이 나중에 aop 사용하게 될 때도 편해요! (userRole이 서비스계층에서 사용될 수 있으니까요)

backswan0 replied:
넵, 확인했습니다! 고민한 부분이었는데, 저희가 AOP 사용도 고려 중이니까 수정하겠습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/keyword/controller/UserKeywordController.java:
```diff
@@ -9,4 +18,20 @@
+    ) {
+        Long userId = authUser.id();
+
+        CreateUserKeywordResponseDto responseDto = userKeywordService.createUserKeyword(
```
전 이걸 return 문에 넣어버렸는데 같이 의논해봐야겠네요

backswan0 replied:
저는 괄호 안에 괄호가 여러 번 들어가서 가독성이 떨어질까 봐 우선 빼두었습니다:) 일단 논의할 사항으로 회의록에 기록해둘게요:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/keyword/service/UserKeywordService.java:
```diff
@@ -1,10 +1,65 @@
+                Keyword foundKeyword = findKeywordById(keywordId);
+
+                // 사용자가 선택한 키워드와 사용자 간의 관계를 나타내는 UserKeyword 엔티티 생성
+                UserKeyword newUserKeyword = UserKeyword.of(
```
이미 해당 사용자가 / 해당 키워드를 구독 중인지 검증하는 로직이 빠졌습니다. 
현재 복합 unique(userId, keywordId) 걸려있어서 이 경우 사용자는 500에러를 받게되겠죠.
이미 구독 중인 키워드는 무시하도록 처리해주세요.

roqkfchqh replied:
++ entity 계층에 추가해야 될 엔티티 검증 로직이라고 생각됩니다!

backswan0 replied:
수정하고 커밋 푸시까지 했습니다:)

roqkfchqh reviewed (changes requested):
수정사항 / 논의사항 커멘트 남겼습니다! 확인하시고 재요청 부탁드립니다!

backswan0 committed:
:memo: chore: 메서드명 수정 및 사용자 id 정제 작업을 컨트롤러에서 서비스 레이어로 이동

- registerUserKeywordList -> createUserKeywordList로 변경

- authUser.id() 정제 작업을 컨트롤러에서 서비스로 이동

- 피드백 3개 중 2개 반영 완료

- [할 일] 해당 사용자가 입력한 키워드로 이미 알림을 받는지 검증 및 예외 처리

llRosell commented:
전체 코드 확인했습니다. 채영님이 말씀하신 부분 외에 다른 comment는 없습니다!

tmdcksdl commented:
저도 전체 코드 확인 완료했습니다.
추가적으로 수정할 부분은 없는 것 같습니다!

backswan0 committed:
:bug: fix: 키워드 중복 추가 검증 로직 추가

- uniqueConstraints 설정 때문에 발생하는 Duplicate entry 오류 해결

- 불필요한 주석 제거

backswan0 committed:
:recycle: refactor: 응답 시 식별자 대신 키워드 이름 반환하도록 로직 수정

- 요청 시 keyword id로 입력되기 때문에 별도의 join문 없이 바로 키워드 조회, 없을 시 not found 반환

- 데이터베이스와 소통하지 않는 메서드는 Keyword 엔티티 내부로 이동

backswan0 committed:
:recycle: refactor: 팀 코드 컨벤션 반영 및 알림 키워드 등록 시 로직 수정

- 불필요한 메서드 삭제 및 리팩토링

- [할 일] user service와 keyword service로 findById 부분 추출

- [할 일] AuthUser를 서비스에서 사용하지 않도록 수정

backswan0 committed:
:recycle: refactor: AuthUser를 service가 아니라 controller에서 사용하고 id 추출

nbcamp-spring commented:
@coderabbitai review

nbcamp-spring commented:
https://github.com/coderabbitai review

nbcamp-spring commented:
@coderabbitai review

nbcamp-spring commented:
@coderabbitai full review

nbcamp-spring commented:
@coderabbitai 응 도와줘

nbcamp-spring commented:
@coderabbitai

src/main/java/com/project/cheerha/domain/keyword/service/UserKeywordService.java (2)
27-40: 사용자 키워드 생성 로직
createUserKeyword는 요청 DTO에서 키워드 ID 목록을 추출 후, 새 키워드를 등록하고 이름 리스트를 반환합니다. 테스트 코드가 TODO로 남아 있으니 추후 보강을 권장합니다.

테스트 코드 작성이 필요하다면 도와드릴까요?

---
너가 여기서 말한 부분을 도와달라는 의미였어
